File: meshplanner-app/src/tests/cross_validation/runner.ts
// @vitest-environment node
/**
 * TypeScript cross-validation runner for ITM path loss.
 *
 * Loads canonical terrain profiles (YAML), runs the JS ITM port
 * (computePathLoss), and compares output to Python golden data
 * from pytest-regressions (YAML).  Reports pass/fail per profile
 * with configurable tolerance (target: ±3 dB).
 *
 * Usage (vitest):
 *   import { runAllValidations } from './runner'
 *   const { passed, failed } = runAllValidations()
 *
 * Usage (standalone):
 *   node --import tsx runner.ts
 */

import { readFileSync, readdirSync } from "node:fs"
import { join, resolve, dirname } from "node:path"
import { fileURLToPath } from "node:url"
import { load } from "js-yaml"
import { computePathLoss } from "../../lib/propagation/itm"
import type { TerrainProfile } from "../../lib/types"

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

/** Repo root — 4 levels up from src/tests/cross_validation/ */
const REPO_ROOT = resolve(__dirname, "../../../../")

/** Directory containing the 10 canonical YAML profile files. */
const CANONICAL_DIR = join(REPO_ROOT, "tests/cross_validation/data/canonical")

/** Directory containing pytest-regressions golden YAML files. */
const GOLDEN_DIR = join(REPO_ROOT, "tests/cross_validation/test_python")

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface CanonicalProfile {
  name: string
  terrain_type: string
  lat1: number
  lon1: number
  lat2: number
  lon2: number
  elevations: number[]
  total_distance_km: number
  frequency_mhz: number
  tx_height_m: number
  rx_height_m: number
  expected_loss_range: [number, number]
}

export interface ValidationResult {
  profile: string
  terrainType: string
  pythonLossDb: number
  jsLossDb: number
  deltaDb: number
  pass: boolean
}

// ---------------------------------------------------------------------------
// Internal types (raw YAML structure)
// ---------------------------------------------------------------------------

interface RawExpectedLoss {
  path_loss_db: number
  path_loss_tolerance: number
  free_space_loss_db: number
  free_space_tolerance: number
  excess_loss_db: number
  excess_loss_tolerance: number
}

interface RawPropagationParams {
  total_distance_km: number
  frequency_mhz: number
  tx_height_m: number
  rx_height_m: number
  polarization: number
  climate: number
  ground_permittivity: number
  ground_conductivity: number
  surface_refractivity: number
}

interface RawEndpoints {
  lat_start: number
  lon_start: number
  lat_end: number
  lon_end: number
}

interface RawProfile {
  version: number
  name: string
  terrain_type: string
  description: string
  endpoints: RawEndpoints
  propagation_params: RawPropagationParams
  expected_loss: RawExpectedLoss
  elevations: number[]
}

interface RawDoc {
  canonical_terrain_profile: RawProfile
}

/** Shape of a pytest-regressions golden data YAML file. */
interface GoldenData {
  profile: string
  path_loss_db: number
  free_space_loss_db: number
  excess_loss_db: number
  distance_km: number
  frequency_mhz: number
}

// ---------------------------------------------------------------------------
// Profile loading
// ---------------------------------------------------------------------------

/**
 * Flatten a raw YAML profile into the simplified CanonicalProfile shape.
 */
function flattenProfile(raw: RawProfile): CanonicalProfile {
  const p = raw.propagation_params
  const e = raw.expected_loss
  return {
    name: raw.name,
    terrain_type: raw.terrain_type,
    lat1: raw.endpoints.lat_start,
    lon1: raw.endpoints.lon_start,
    lat2: raw.endpoints.lat_end,
    lon2: raw.endpoints.lon_end,
    elevations: raw.elevations,
    total_distance_km: p.total_distance_km,
    frequency_mhz: p.frequency_mhz,
    tx_height_m: p.tx_height_m,
    rx_height_m: p.rx_height_m,
    expected_loss_range: [e.path_loss_db, e.path_loss_tolerance],
  }
}

/**
 * Load all canonical terrain profiles from disk.
 *
 * Reads every `*.yaml` file under `CANONICAL_DIR`, parses it, and
 * returns a sorted array of flattened `CanonicalProfile` objects.
 */
export function loadProfiles(): CanonicalProfile[] {
  const files = readdirSync(CANONICAL_DIR)
    .filter((f) => f.endsWith(".yaml"))
    .sort()

  const profiles: CanonicalProfile[] = []
  for (const file of files) {
    const raw = readFileSync(join(CANONICAL_DIR, file), "utf-8")
    const doc = load(raw) as RawDoc
    profiles.push(flattenProfile(doc.canonical_terrain_profile))
  }
  return profiles
}

// ---------------------------------------------------------------------------
// Golden data loading
// ---------------------------------------------------------------------------

/**
 * Load the Python golden path_loss_db for a given profile name.
 *
 * The golden data was generated by `pytest-regressions` and stored as
 * a flat YAML file at `tests/cross_validation/test_python/<name>.yml`.
 *
 * @param profileName — profile name (e.g. "flat", "mountain_ridge")
 * @returns The Python reference `path_loss_db` value.
 */
export function loadGoldenData(profileName: string): number {
  const filePath = join(GOLDEN_DIR, `${profileName}.yml`)
  const raw = readFileSync(filePath, "utf-8")
  const golden = load(raw) as GoldenData
  return golden.path_loss_db
}

// ---------------------------------------------------------------------------
// Profile conversion
// ---------------------------------------------------------------------------

/**
 * Convert a CanonicalProfile to the TerrainProfile type expected by
 * computePathLoss.
 */
function toTerrainProfile(p: CanonicalProfile): TerrainProfile {
  const n = p.elevations.length
  const elevations = new Float64Array(p.elevations)
  const distancesKm = new Float64Array(n)
  for (let i = 0; i < n; i++) {
    distancesKm[i] = (p.total_distance_km * i) / (n - 1)
  }

  let maxElev = -Infinity
  let minElev = Infinity
  let sum = 0
  for (const e of p.elevations) {
    if (e > maxElev) maxElev = e
    if (e < minElev) minElev = e
    sum += e
  }

  return {
    elevations,
    distancesKm,
    totalDistanceKm: p.total_distance_km,
    maxElevation: maxElev,
    minElevation: minElev,
    avgElevation: sum / n,
    latlons: [],
  }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

/**
 * Validate a single canonical profile against Python golden data.
 *
 * Runs the JS `computePathLoss` for the given profile, loads the
 * corresponding Python golden value, and compares the two.
 *
 * @param profile — the canonical profile to validate
 * @returns A ValidationResult with JS/Python values and pass/fail verdict.
 */
export function validateProfile(profile: CanonicalProfile): ValidationResult {
  const terrain = toTerrainProfile(profile)
  const result = computePathLoss(terrain, {
    frequencyMhz: profile.frequency_mhz,
    txHeightM: profile.tx_height_m,
    rxHeightM: profile.rx_height_m,
  })

  const pyLoss = loadGoldenData(profile.name)
  const deltaDb = Math.round((result.pathLossDb - pyLoss) * 100) / 100

  return {
    profile: profile.name,
    terrainType: profile.terrain_type,
    pythonLossDb: pyLoss,
    jsLossDb: result.pathLossDb,
    deltaDb,
    pass: false, // caller sets pass/fail based on tolerance
  }
}

// ---------------------------------------------------------------------------
// Batch runner
// ---------------------------------------------------------------------------

/**
 * Run cross-validation for all canonical profiles.
 *
 * @param toleranceDb — maximum acceptable absolute deviation in dB (default 3.0)
 * @returns Aggregated results including pass/fail counts.
 */
export function runAllValidations(
  toleranceDb: number = 3.0,
): {
  results: ValidationResult[]
  passed: number
  failed: number
  total: number
} {
  const profiles = loadProfiles()
  const results: ValidationResult[] = []

  for (const p of profiles) {
    const vr = validateProfile(p)
    vr.pass = Math.abs(vr.deltaDb) <= toleranceDb
    results.push(vr)
  }

  let passed = 0
  let failed = 0
  for (const r of results) {
    if (r.pass) passed++
    else failed++
  }

  return { results, passed, failed, total: results.length }
}

// ---------------------------------------------------------------------------
// Standalone entry point
// ---------------------------------------------------------------------------

if (process.argv[1] && fileURLToPath(import.meta.url) === resolve(process.argv[1])) {
  const maxTolerance = parseFloat(process.argv[2] ?? "3")
  const { results, passed, failed, total } = runAllValidations(maxTolerance)

  const report = {
    engine: "js (full ITM / Longley-Rice)",
    tolerance_db: maxTolerance,
    generated_at: new Date().toISOString(),
    profiles_loaded: total,
    summary: { passed, failed, total },
    results: results.map((r) => ({
      profile: r.profile,
      terrain_type: r.terrainType,
      python_loss_db: r.pythonLossDb,
      js_loss_db: r.jsLossDb,
      delta_db: r.deltaDb,
      pass: r.pass,
    })),
  }

  process.stdout.write(`${JSON.stringify(report, null, 2)}\n`)

  if (failed > 0) {
    process.exit(1)
  }
}

File: tests/cross_validation/runner.py
#!/usr/bin/env python3
"""
ITM Cross-Validation Runner.

Loads each canonical terrain profile, computes the Python ITM golden value,
invokes the JS ITM engine via Node/tsx, and reports the comparison.

Usage:
    python tests/cross_validation/runner.py                    # full report
    python tests/cross_validation/runner.py --json             # JSON output
    python tests/cross_validation/runner.py --fail-fast        # exit on first diff > tolerance
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

import yaml

from meshplanner.propagation.itm import compute_path_loss

REPO_ROOT = Path(__file__).resolve().parents[2]
PROFILES_DIR = REPO_ROOT / "tests" / "cross_validation" / "data" / "canonical"
JS_TEST_SCRIPT = (
    REPO_ROOT
    / "meshplanner-app"
    / "src"
    / "tests"
    / "cross_validation"
    / "test_js.test.ts"
)

# Colours for terminal output
_GREEN = "\033[92m"
_RED = "\033[91m"
_YELLOW = "\033[93m"
_CYAN = "\033[96m"
_BOLD = "\033[1m"
_RESET = "\033[0m"


def _load_profiles() -> list[dict]:
    files = sorted(PROFILES_DIR.glob("*.yaml"))
    profiles = []
    for f in files:
        with open(f) as fh:
            doc = yaml.safe_load(fh)
        profiles.append(doc["canonical_terrain_profile"])
    return profiles


def _run_python(profile: dict) -> dict:
    """Run Python ITM for a profile and return full result dict."""
    params = profile["propagation_params"]
    return compute_path_loss(
        elevations=profile["elevations"],
        total_distance_km=params["total_distance_km"],
        frequency_mhz=params["frequency_mhz"],
        tx_height_m=params["tx_height_m"],
        rx_height_m=params["rx_height_m"],
        polarization=params["polarization"],
        climate=params["climate"],
        ground_permittivity=params["ground_permittivity"],
        ground_conductivity=params["ground_conductivity"],
        surface_refractivity=params["surface_refractivity"],
    )


def _run_js() -> list[dict]:
    """Invoke Node/tsx to run the JS ITM and return parsed results."""
    result = subprocess.run(
        ["npx", "tsx", str(JS_TEST_SCRIPT)],
        capture_output=True,
        text=True,
        cwd=REPO_ROOT / "meshplanner-app",
        timeout=60,
    )
    if result.returncode != 0:
        print(f"STDERR: {result.stderr}", file=sys.stderr)
        result.check_returncode()
    data = json.loads(result.stdout)
    return data["results"]


def _compare(
    py_result: dict,
    js_entry: dict,
    expected: dict,
) -> dict:
    """Compare Python and JS results for a single profile."""
    pl_dev = js_entry["js_path_loss_db"] - py_result["path_loss_db"]
    fs_dev = js_entry["js_free_space_loss_db"] - py_result["free_space_loss_db"]
    ex_dev = js_entry["js_excess_loss_db"] - py_result["excess_loss_db"]

    pl_tol = expected["path_loss_tolerance"]
    fs_tol = expected["free_space_tolerance"]
    ex_tol = expected["excess_loss_tolerance"]

    return {
        "profile": js_entry["profile"],
        "python": {
            "path_loss_db": py_result["path_loss_db"],
            "free_space_loss_db": py_result["free_space_loss_db"],
            "excess_loss_db": py_result["excess_loss_db"],
        },
        "javascript": {
            "path_loss_db": js_entry["js_path_loss_db"],
            "free_space_loss_db": js_entry["js_free_space_loss_db"],
            "excess_loss_db": js_entry["js_excess_loss_db"],
        },
        "deviation": {
            "path_loss_db": round(pl_dev, 1),
            "free_space_loss_db": round(fs_dev, 1),
            "excess_loss_db": round(ex_dev, 1),
        },
        "within_tolerance": {
            "path_loss_db": abs(pl_dev) <= pl_tol,
            "free_space_loss_db": abs(fs_dev) <= fs_tol,
            "excess_loss_db": abs(ex_dev) <= ex_tol,
        },
    }


def main():
    args = set(sys.argv[1:])
    as_json = "--json" in args
    fail_fast = "--fail-fast" in args

    profiles = _load_profiles()

    print(
        f"{_BOLD}ITM Cross-Validation Runner{_RESET}\n"
        f"  Engine:     Python (itmlogic) vs JS (knife-edge diffraction)\n"
        f"  Profiles:   {len(profiles)}\n"
    )

    # Run Python ITM on all profiles
    py_results = {}
    for p in profiles:
        py_results[p["name"]] = _run_python(p)

    # Run JS ITM via Node/tsx
    js_results_list = _run_js()
    js_results = {r["profile"]: r for r in js_results_list}

    # Compare
    comparisons = []
    for p in profiles:
        name = p["name"]
        py = py_results[name]
        js = js_results[name]
        cmp = _compare(py, js, p["expected_loss"])
        comparisons.append(cmp)

    if as_json:
        print(json.dumps(comparisons, indent=2))
        return

    # Terminal report
    total = len(comparisons)
    passed = 0
    failed = 0

    for cmp in comparisons:
        wt = cmp["within_tolerance"]
        all_ok = wt["path_loss_db"] and wt["free_space_loss_db"] and wt["excess_loss_db"]
        status = f"{_GREEN}PASS{_RESET}" if all_ok else f"{_RED}FAIL{_RESET}"
        icon = "✓" if all_ok else "✗"

        if all_ok:
            passed += 1
        else:
            failed += 1
            if fail_fast:
                print(f"\n{_RED}Failing fast on {cmp['profile']}{_RESET}")

        d = cmp["deviation"]
        wt_labels = {
            k: f"{'✓' if v else '✗'} {k.replace('_', ' ')}"
            for k, v in wt.items()
        }
        print(
            f"  {icon} {_BOLD}{cmp['profile']:25s}{_RESET}  {status}\n"
            f"     Python  PL={cmp['python']['path_loss_db']:6.1f} dB  "
            f"FS={cmp['python']['free_space_loss_db']:6.1f} dB  "
            f"EX={cmp['python']['excess_loss_db']:5.1f} dB\n"
            f"     JS      PL={cmp['javascript']['path_loss_db']:6.1f} dB  "
            f"FS={cmp['javascript']['free_space_loss_db']:6.1f} dB  "
            f"EX={cmp['javascript']['excess_loss_db']:5.1f} dB\n"
            f"     Δ       PL={d['path_loss_db']:6.1f} dB  "
            f"FS={d['free_space_loss_db']:6.1f} dB  "
            f"EX={d['excess_loss_db']:5.1f} dB\n"
            f"     Tol:    {wt_labels['path_loss_db']}, "
            f"{wt_labels['free_space_loss_db']}, "
            f"{wt_labels['excess_loss_db']}"
        )

    print(
        f"\n{'=' * 60}\n"
        f"  Results: {_GREEN}{passed} passed{_RESET}, "
        f"{_RED}{failed} failed{_RESET} / {total} total\n"
        f"  Note: JS uses a simplified knife-edge model; "
        f"mismatches for complex terrain are expected.\n"
        f"  For golden-data regression tests, run:\n"
        f"    pytest tests/cross_validation/   (Python)\n"
        f"    npx vitest run --project cross-validation  (JS)\n"
    )

    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    main()
